File: sdk/android/story-sdk/src/main/kotlin/com/openstory/sdk/internal/ui/StoryPlaybackProgressState.kt
package com.openstory.sdk.internal.ui

internal data class StoryPlaybackProgressState(
    val totalDurationMs: Long,
    val remainingDurationMs: Long,
) {
    fun fractionCompleted(): Float {
        val elapsedDurationMs = (totalDurationMs - remainingDurationMs).coerceIn(0L, totalDurationMs)
        return (elapsedDurationMs.toFloat() / totalDurationMs.toFloat()).coerceIn(0f, 1f)
    }

    fun afterElapsed(elapsedDurationMs: Long): StoryPlaybackProgressState {
        return copy(
            remainingDurationMs = (remainingDurationMs - elapsedDurationMs).coerceIn(0L, totalDurationMs),
        )
    }

    companion object {
        fun started(totalDurationMs: Long): StoryPlaybackProgressState {
            val normalizedDurationMs = totalDurationMs.coerceAtLeast(1L)
            return StoryPlaybackProgressState(
                totalDurationMs = normalizedDurationMs,
                remainingDurationMs = normalizedDurationMs,
            )
        }
    }
}

File: sdk/flutter/ios/OpenStorySDK/Internal/UI/StoryPlaybackProgressState.swift
import Foundation

internal struct StoryPlaybackProgressState: Equatable {
    let totalDurationMs: Int64
    let remainingDurationMs: Int64

    func fractionCompleted() -> CGFloat {
        let elapsed = max(0, min(totalDurationMs, totalDurationMs - remainingDurationMs))
        guard totalDurationMs > 0 else {
            return 1
        }
        return CGFloat(Double(elapsed) / Double(totalDurationMs))
    }

    func afterElapsed(_ elapsedDurationMs: Int64) -> StoryPlaybackProgressState {
        StoryPlaybackProgressState(
            totalDurationMs: totalDurationMs,
            remainingDurationMs: max(0, min(totalDurationMs, remainingDurationMs - elapsedDurationMs))
        )
    }

    static func started(_ totalDurationMs: Int64) -> StoryPlaybackProgressState {
        let normalizedDurationMs = max(1, totalDurationMs)
        return StoryPlaybackProgressState(
            totalDurationMs: normalizedDurationMs,
            remainingDurationMs: normalizedDurationMs
        )
    }
}
